server/src/main/java/org/elasticsearch/script/VectorScoreScriptUtils.java

```diff
@@ -307,6 +307,84 @@ public interface DotProductInterface {
         double dotProduct();
     }
 
+    public static class BitDotProduct extends DenseVectorFunction implements DotProductInterface {
+        private final byte[] byteQueryVector;
+        private final float[] floatQueryVector;
+
+        public BitDotProduct(ScoreScript scoreScript, DenseVectorDocValuesField field, byte[] queryVector) {
+            super(scoreScript, field);
+            if (field.getElementType() != DenseVectorFieldMapper.ElementType.BIT) {
+                throw new IllegalArgumentException("cannot calculate bit dot product for non-bit vectors");
+            }
+            int fieldDims = field.get().getDims();
+            if (fieldDims != queryVector.length * Byte.SIZE && fieldDims != queryVector.length) {
+                throw new IllegalArgumentException(
+                    "The query vector has an incorrect number of dimensions. Must be ["
+                        + fieldDims / 8
+                        + "] for bitwise operations, or ["
+                        + fieldDims
+                        + "] for byte wise operations: provided ["
+                        + queryVector.length
+                        + "]."
+                );
+            }
+            this.byteQueryVector = queryVector;
+            this.floatQueryVector = null;
+        }
+
+        public BitDotProduct(ScoreScript scoreScript, DenseVectorDocValuesField field, List<Number> queryVector) {
+            super(scoreScript, field);
+            if (field.getElementType() != DenseVectorFieldMapper.ElementType.BIT) {
+                throw new IllegalArgumentException("cannot calculate bit dot product for non-bit vectors");
+            }
+            float[] floatQueryVector = new float[queryVector.size()];
+            byte[] byteQueryVector = new byte[queryVector.size()];
+            boolean isFloat = false;
+            for (int i = 0; i < queryVector.size(); i++) {
+                Number number = queryVector.get(i);
+                floatQueryVector[i] = number.floatValue();
+                byteQueryVector[i] = number.byteValue();
+                if (floatQueryVector[i] % 1.0f != 0.0f || floatQueryVector[i] < Byte.MIN_VALUE || floatQueryVector[i] > Byte.MAX_VALUE) {
+                    isFloat = true;
+                }
```

Comment:
I think we can `break` here

Reply:
We need to build both vectors and then pick the right now, breaking would prevent us from building the arrays.